platform-hub-api/app/controllers/concerns/authentication.rb

```diff
@@ -0,0 +1,50 @@
+module Authentication
+  extend ActiveSupport::Concern
+
+  include ActionController::HttpAuthentication::Token
+
+  def require_authentication
+    head :unauthorized unless authenticated?
+  end
+
+  def authenticated?
+    current_user.present?
+  end
+
+  def current_user
+    token, _ = token_and_options(request)
+
+    return nil if token.blank?
+
+    @current_user ||= create_or_fetch_user(token)
+  end
+
+  private
+
+  def create_or_fetch_user token
+    # Important assumptions about the token:
+    # - it is already verified by the auth proxy in front
+    # - expiry has already been checked by the auth proxy in front (TODO: perhaps recheck here)
+    begin
+      payload, _ = JWT.decode(token, nil, false)
+    rescue => e
+      logger.error "Failed to parse JWT token: #{token}. Error: #{e.message}"
+      return nil
+    end
+
+    return nil if payload.blank? || payload['sub'].blank?
+
+    id = payload['sub']
+
+    user = User.find_by_id(id)
+
+    if user.blank?
+      user = NewUserService.new.create payload
```

Comment:
> (albeit this particular service is stateless and behaves in a highly functional way) 

+1 for functional. At the moment the only thing you want from this services is to create a new user and the service itself is stateless so no point in creating an instance of it every time. I'm not fussed at this stage tho. Can be easily changed moving forward.